test/unit/org/apache/cassandra/service/accord/EpochSyncTest.java

```diff
@@ -0,0 +1,673 @@
+                                                (fn, time, unit) -> {},
+                                                failures::add);
+        }
+
+        void addNode(Node.Id id, long token)
+        {
+            Invariants.checkState(!tokens.contains(token), "Attempted to add token %d for node %s but token is already taken", token, id);
+            Epoch epoch = Epoch.create(current.epoch.getEpoch() + 1);
+
+            Instance instance = new Instance(id, token, epoch, createMessaging(id), fd);
+            instances.put(id, instance);
+            tokens.add(token);
+
+            current = current.forceEpoch(epoch)
+                             .withPlacements(DataPlacements.builder(2)
+                                                           .with(meta, DataPlacement.empty())
+                                                           .with(replication_params, rebuildPlacements(epoch))
+                                                           .build())
+                             .withDirectory(current.directory.with(new NodeAddresses(address(id)), new Location("dc1", "r1")));
+            notify(current);
+        }
+
+        void removeNode(Node.Id pick)
+        {
+            Instance inst = Objects.requireNonNull(instances.get(pick), "Unknown id " + pick);
+            Invariants.checkState(!removed.contains(pick), "Can not remove node twice; node " + pick);
+            tokens.remove(inst.token);
+            removed.add(pick);
+            current = current.forceEpoch(Epoch.create(current.epoch.getEpoch() + 1))
+                             .withDirectory(current.directory.without(new NodeId(pick.id)));
+
+            current = current.withPlacements(DataPlacements.builder(2)
+                                                           .with(meta, DataPlacement.empty())
+                                                           .with(replication_params, rebuildPlacements(current.epoch))
+                                                           .build());
+            notify(current);
+        }
+
+        private DataPlacement rebuildPlacements(Epoch epoch)
+        {
+            DataPlacement.Builder builder = DataPlacement.builder();
+            for (Node.Id inst : alive())
+                for (Replica replica : instances.get(inst).replica())
+                    builder.withReadReplica(epoch, replica).withWriteReplica(epoch, replica);
+            return builder.build();
+        }
+
+        void bumpEpoch()
+        {
+            current = current.forceEpoch(Epoch.create(current.epoch.getEpoch() + 1));
+            notify(current);
+        }
+
+        private void notify(ClusterMetadata current)
+        {
+            Ranges ranges = AccordTopology.createAccordTopology(current).ranges().mergeTouching();
+            if (!current.directory.isEmpty())
+                Assertions.assertThat(ranges).hasSize(1);
```

Comment:
there was a bug early on where there were gaps in the token ring... this isn't *needed* anymore, but left just to make sure nothing breaks again